File: programmers/algorithm_data_structure/greedy/구명보트.py
from collections import deque


def solution(people, limit):
    '''
    매번 가장 무거운 사람을 1명 넣고
    가장 가벼운 사람 순으로 나머지 무게를 채워 넣는다.

    deque를 이용하여 양쪽에서 빼는 시간을 최소화
    (다른 사람 코드를 보면 index 변수 2개를 만들어 간단하게 처리)
    '''
    sorted_people = deque(sorted(people))
    answer = 0
    while sorted_people:
        boat = []
        while sorted_people and sum(boat) + sorted_people[0] <= limit:
            boat.append(sorted_people.popleft())

        answer += 1

    return answer

File: programmers/algorithm_data_structure/greedy/구명보트.java
package algorithm_data_structure.greedy;

import java.util.Arrays;

class 구명보트 {

    /*
     * 매번 제일 무거운 사람과 가벼운 사람을 태워 보낸다.
     */
    public int solution(int[] people, int limit) {
        int answer = 0;
        int lightPerson = 0;
        int heavyPerson = people.length - 1;
        int[] sortedPeople = Arrays.copyOf(people, people.length);
        Arrays.sort(sortedPeople);

        while (lightPerson < heavyPerson) {
            if (sortedPeople[lightPerson] + sortedPeople[heavyPerson] <= limit) {
                lightPerson += 1;
                heavyPerson -= 1;
            }else{
                heavyPerson -= 1;
            }
            answer += 1;
        }

        // 한 사람만 남은 경우
        if (lightPerson == heavyPerson){
            answer += 1;
        }

        return answer;
    }
}